src/gateway/routers/RouterAggregator.sol

```diff
@@ -0,0 +1,231 @@
+    event ExecuteMessagRecovery(bytes message);
+    event File(bytes32 indexed what, address[] routers);
+
+    constructor(address gateway_) {
+        gateway = GatewayLike(gateway_);
+
+        wards[msg.sender] = 1;
+        emit Rely(msg.sender);
+    }
+
+    // --- Administration ---
+    function file(bytes32 what, address[] calldata routers_) external auth {
+        if (what == "routers") {
+            require(routers_.length <= MAX_ROUTER_COUNT, "RouterAggregator/exceeds-max-router-count");
+
+            // Enable new routers, setting quorum to number of routers
+            uint8 quorum_ = uint8(routers_.length);
+            for (uint8 i; i < routers_.length; ++i) {
+                // Ids are assigned sequentially starting at 1
+                validRouters[routers_[i]] = Router(i + 1, quorum_);
+            }
+
+            // Disable old routers that weren't already overridden
+            for (uint8 j = uint8(routers_.length); j < routers.length; ++j) {
+                delete validRouters[address(routers[j])];
+            }
+
+            routers = routers_;
+        } else {
+            revert("RouterAggregator/file-unrecognized-param");
+        }
+
+        emit File(what, routers_);
+    }
+
+    // --- Incoming ---
+    /// @dev Handle incoming messages, proofs, and recoveries.
+    ///      Assumes routers ensure messages cannot be confirmed more than once.
+    function handle(bytes calldata payload) public {
+        Router memory router = validRouters[msg.sender];
+        require(router.id != 0, "RouterAggregator/invalid-router");
+        _handle(payload, router);
+    }
+
+    function _handle(bytes calldata payload, Router memory router) internal {
+        if (MessagesLib.isRecoveryMessage(payload)) {
+            require(routers.length > 1, "RouterAggregator/no-recovery-with-one-router-allowed");
+            return _handleRecovery(payload);
+        }
+
+        if (router.quorum == 1 && !MessagesLib.isMessageProof(payload)) {
+            // Special case for gas efficiency
+            gateway.handle(payload);
+            emit ExecuteMessage(payload, msg.sender);
+            return;
+        }
+
+        bytes32 messageHash;
+        ConfirmationState storage state;
+        if (MessagesLib.isMessageProof(payload)) {
+            messageHash = MessagesLib.parseMessageProof(payload);
+            state = _confirmations[messageHash];
+            state.proofs[router.id - 1]++;
+
+            emit HandleProof(messageHash, msg.sender);
+        } else {
+            messageHash = keccak256(payload);
+            state = _confirmations[messageHash];
+            state.messages[router.id - 1]++;
+
+            emit HandleMessage(payload, msg.sender);
+        }
+
+        if (state.messages.countNonZeroValues() >= 1 && state.proofs.countNonZeroValues() >= router.quorum - 1) {
+            // Reduce total message confiration count by 1, by finding the first non-zero value
```

Comment:
`// Reduce total message confirmation count by 1, by finding the first non-zero value`